### PR TITLE
[CR/WH] [000000] refactor strategies approach

### PR DESCRIPTION
**Pull request by colinrymer**

### CODE REVIEW

File: lib/phenotype/versioned_app.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      strategies.push(strat)
+    end
+
+    def add_version(version:, &block)
+      routes[version] = block
+    end
+
+    def call(env)
+      strats = strategies.find_all { |strat| strat.version(env) }
+      return no_strategy_response.call(env) if strategies.empty?
```
Comment (wshatch) [inline lib/phenotype/versioned_app.rb:21]: I just realized, how are we going to handle versioning with apps where we can't necessarily change the client namely mobile API and old versions of our mobile apps?

Comment (pasha) [inline lib/phenotype/versioned_app.rb:21]: :+1:

Comment (colinrymer) [inline lib/phenotype/versioned_app.rb:21]: That'll be handled at the API gateway layer.

Comment (pasha) [inline lib/phenotype/versioned_app.rb:21]: Do you have any recommendations on how to handle this case https://github.com/rentpath/metasaurus/pull/259  I was thinking to adding versioning to Metasaurus then merging this PR #259 as v2, we can then slowly release apps depending on newer version.

Comment (pasha) [inline lib/phenotype/versioned_app.rb:21]: @colinrymer thoughts?

Comment (colinrymer) [inline lib/phenotype/versioned_app.rb:21]: A potential solution would be a separate middleware that checks for the presence of a version and adds it if not found. That keeps this library nice and clean and doesn't suggest to future users that providing a default is something that should normally be done.

@wshatch is out until tomorrow, but I'd like to wait to get his thoughts on this.

Comment (wshatch) [inline lib/phenotype/versioned_app.rb:21]: So if we have a 406 status code, the default middleware would execute the default route? That should work. We could also give it a silly, somewhat shameful name like http://ruby-doc.org/stdlib-1.9.3/libdoc/minitest/unit/rdoc/MiniTest/Unit/TestCase.html#method-c-i_suck_and_my_tests_are_order_dependent-21

Comment (colinrymer) [inline lib/phenotype/versioned_app.rb:21]: My thought was that the default middleware would just add a version header.

**Comment by pasha** [comment]
👍

**Pull request merged**